### PR TITLE
Determine latest upstream version directly

### LINKED ISSUE #70: Determine latest upstream version to upgrade directly
Right now `upgrade-provider` determines the latest upstream version by looking at all open issues with a specified title opened by our pulumi-bot. Rather than require external provider maintainers to set up a similar mechanism, we should fetch the latest version directly

### PR DESCRIPTION
**Pull request by aq17**
Fixes #70 

We should be able to fetch the latest upstream provider version to avoid users having to use a similar zapier/ pulumi-bot system of creating `Upgrade terraform-provider-...` issues

### CODE REVIEW

File: upgrade/steps-helpers.go [hunk 1]
```diff
@@ -351,3 +352,113 @@ func getRepoExpectedLocation(ctx Context, cwd, repoPath string) (string, error)
...
+		Type  string `json:"TYPE"`
+	}{}
+	err = json.Unmarshal(bytes.Bytes(), &release)
+	if err != nil {
+		return nil, "", err
+	}
+
+	contract.Assert(len(release) == 1)
+	contract.Assert(release[0].Type == "Latest")
+	v, err := semver.NewVersion(release[0].Title)
```
Comment (aq17) [inline upgrade/steps-helpers.go:393]: Trying to figure out how to unmarshal the output – currently getting invalid character 'v' looking for beginning of value.
The output of bytes.String() is v4.120.0       Latest  v4.120.0        2023-05-10T17:30:24Z.

Running gh release list --repo=databricks/terraform-provider-databricks --exclude-drafts --exclude-pre-releases -L=1 gives:
TITLE TYPE TAG NAME PUBLISHED
v1.15.0 Latest v1.15.0 about 5 days ago

File: upgrade/steps-helpers.go [hunk 2]
```diff
@@ -9,10 +9,11 @@ import (
 	"os/exec"
 	"path/filepath"
 	"regexp"
+	"sort"
 	"strings"
 
 	"github.com/Masterminds/semver/v3"
-	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	"github.com/pulumi/pulumi/sdk/go/common/util/contract"
```
Comment (iwahbe) [inline upgrade/steps-helpers.go:16]: Is this change on purpose?

File: upgrade/steps-helpers.go [hunk 3]
```diff
@@ -351,3 +352,113 @@ func getRepoExpectedLocation(ctx Context, cwd, repoPath string) (string, error)
...
 }
+
+// Fetch the expected upgrade target from github. Return a list of open upgrade issues,
+// sorted by semantic version. The list may be empty.
+//
+// The second argument represents a message to describe the result. It may be empty.
+func GetExpectedTarget(ctx Context, name, upstreamOrg string) (*UpstreamUpgradeTarget, string, error) {
+	if ctx.TargetVersion != nil {
+		return &UpstreamUpgradeTarget{Version: ctx.TargetVersion}, "", nil
+	}
```
Comment (iwahbe) [inline upgrade/steps-helpers.go:363]: If `TargetVersion != nil && ctx.InferVersion`, we should replicate the behavior of our old `--max-version` flag.

File: upgrade/steps-helpers.go [hunk 4]
```diff
@@ -351,3 +352,113 @@ func getRepoExpectedLocation(ctx Context, cwd, repoPath string) (string, error)
...
+	if !ctx.InferVersion {
+		return getExpectedTargetLatest(ctx, name, upstreamOrg)
+	}
+	return getExpectedTargetFromIssues(ctx, name)
+}
+
+func getExpectedTargetLatest(ctx Context, name, upstreamOrg string) (*UpstreamUpgradeTarget, string, error) {
+	latest := exec.CommandContext(ctx, "gh", "release", "list",
+		"--repo="+upstreamOrg+"/"+ctx.UpstreamProviderName,
+		"-L=1",
```
Comment (iwahbe) [inline upgrade/steps-helpers.go:373]: Long arguments are clearer.
```suggestion
		"--limit=1",
```

File: upgrade/steps-helpers.go [hunk 5]
```diff
@@ -351,3 +352,113 @@ func getRepoExpectedLocation(ctx Context, cwd, repoPath string) (string, error)
...
+	latest.Stdout = bytes
+	err := latest.Run()
+	if err != nil {
+		return nil, "", err
+	}
+	release := []struct {
+		Title string `json:"TITLE"`
+		Type  string `json:"TYPE"`
+	}{}
+	err = json.Unmarshal(bytes.Bytes(), &release)
```
Comment (iwahbe) [inline upgrade/steps-helpers.go:386]: How are you getting JSON from ` gh release list --repo=pulumi/pulumi-aws -L=1 --exclude-drafts --exclude-pre-releases`. I'm just getting text such as:

```
v5.40.0	Latest	v5.40.0	2023-05-06T01:48:33Z
```

File: upgrade/steps-helpers.go (hunk 1 above)
Comment (iwahbe) [reply upgrade/steps-helpers.go:393]: What is the value of `release[0].Title`?

File: upgrade/steps-helpers.go [hunk 6]
```diff
@@ -351,3 +352,113 @@ func getRepoExpectedLocation(ctx Context, cwd, repoPath string) (string, error)
...
+	contract.Assert(len(release) == 1)
+	contract.Assert(release[0].Type == "Latest")
+	v, err := semver.NewVersion(release[0].Title)
+	if err != nil {
+		return nil, "", err
+	}
+	return &UpstreamUpgradeTarget{Version: v}, "", nil
+}
+
+func getExpectedTargetFromIssues(ctx Context, name string) (*UpstreamUpgradeTarget, string, error) {
```
Comment (iwahbe) [inline upgrade/steps-helpers.go:400]: Is this the same as what `getExpectedTarget` used to be?

File: upgrade/steps-helpers.go (hunk 5 above)
Comment (aq17) [reply upgrade/steps-helpers.go:386]: I had assumed there was a flag/ option to return the text as JSON, but that doesn't look like the case so actually shouldn't be using `json.Unmarshal()`

File: upgrade/steps-helpers.go (hunk 6 above)
Comment (aq17) [reply upgrade/steps-helpers.go:400]: Yes, now `GetExpectedTarget` calls either `getExpectedTargetLatest` or `getExpectedTargetFromIssues` depending on the flags set

File: upgrade/steps-helpers.go (hunk 1 above)
Comment (aq17) [reply upgrade/steps-helpers.go]: Empty 😭

**Comment by aq17** [comment]
PR created with `--pulumi-infer-version=false` https://github.com/pulumi/pulumi-oci/pull/131

**Review by iwahbe — APPROVED** [review]
This looks good to me.
